zson/marshal.go

```diff
@@ -514,27 +519,66 @@ func (m *MarshalZNGContext) encodeArrayBytes(arrayVal reflect.Value) (zed.Type,
 }
 
 func (m *MarshalZNGContext) encodeArray(arrayVal reflect.Value) (zed.Type, error) {
-	len := arrayVal.Len()
+	push := m.Builder
+	m.Builder = *zcode.NewBuilder()
 	m.Builder.BeginContainer()
-	var innerType zed.Type
-	for i := 0; i < len; i++ {
+	arrayLen := arrayVal.Len()
+	types := make([]zed.Type, 0, arrayLen)
+	typeMap := make(map[zed.Type]struct{})
+	for i := 0; i < arrayLen; i++ {
 		item := arrayVal.Index(i)
 		typ, err := m.encodeValue(item)
 		if err != nil {
 			return nil, err
 		}
-		// XXX See bug #2575
-		innerType = typ
+		types = append(types, typ)
+		typeMap[typ] = struct{}{}
 	}
 	m.Builder.EndContainer()
-	if innerType == nil {
+	var innerType zed.Type
+	switch len(typeMap) {
+	case 0:
 		// if slice was empty, look up the type without a value
 		var err error
 		innerType, err = m.lookupType(arrayVal.Type().Elem())
 		if err != nil {
 			return nil, err
 		}
+		push.Append(m.Builder.Bytes().Body())
+	case 1:
+		innerType = types[0]
+		push.Append(m.Builder.Bytes().Body())
+	default:
+		unionTypes := make([]zed.Type, 0, len(typeMap))
+		for typ := range typeMap {
+			unionTypes = append(unionTypes, typ)
+		}
+		unionType := m.Context.LookupTypeUnion(unionTypes)
+		// Now we know all the types and have the union type
+		// so we can recode the array with tagged union elements.
+		// We throw out the array computed above and start over with
+		// an empty builder.
+		m.Builder = *zcode.NewBuilder()
+		m.Builder.Truncate()
```

Comment:
No need for a new builder; Truncate's sufficient.
```suggestion
		m.Builder.Truncate()
```